internal/copy_handler.go

```diff
@@ -33,50 +32,59 @@ func HandleCopy(fromConfigFile string, toConfigFile string, backupName string, w
 }
 
 func StartCopy(infos []CopyingInfo) (bool, error) {
-	var maxParallelJobsCount = 8 // TODO place for improvement
-	for i := 0; i < len(infos); i += maxParallelJobsCount {
-		errors := make(chan error)
-		wgDone := make(chan bool)
-
-		var lastIndex = utility.Min(i+maxParallelJobsCount, len(infos))
-		var infosToCopy = infos[i:lastIndex]
-		var wg sync.WaitGroup
-		for _, info := range infosToCopy {
-			wg.Add(1)
-			go copyObject(info, &wg, errors)
-		}
-		wg.Wait()
-		close(wgDone)
-
-		select {
-		case <-wgDone:
-			break
-		case err := <-errors:
-			close(errors)
-			return false, err
+	maxParallelJobsCount := 8
+
+	tickets := make(chan interface{}, maxParallelJobsCount)
+
+	for t := 0; t < maxParallelJobsCount; t++ {
+		tickets <- nil
+	}
+
+	errors := make(chan error, maxParallelJobsCount * 2)
+	var wg sync.WaitGroup
+
+	for _, info := range infos {
+
+		// do we have any errs yet?
+		for len(errors) > 0 {
+			if err := <-errors; err != nil {
+				return false, err
+			}
 		}
+
+		// block here
+		_ = <-tickets
+		wg.Add(1)
+
+		go func(info CopyingInfo) {
+			defer wg.Done()
+			err := copyObject(info)
+			tickets <- nil
+			errors <- err
+		}(info)
 	}
+
+	wg.Wait()
```

Comment:
It looks like some errors are ignored here

Reply:
sure, fixed